### PR TITLE
f-searchbox@v4.0.0 beta.23

### PR DESCRIPTION
**Pull request by kevinrodrigues**
### Added
- Clear button to Loqate feature so addresses can be cleared.
- Refactor loading indicator component so it's shared.
- Added tests.

### Video
https://www.loom.com/share/94dfc3783bd249d9b8997d1aac955d2e

### CODE REVIEW

File: packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<template>
+    <div
+        v-if="shouldDisplayLoadingIndicator"
+        data-test-id="loading-spinner-indicator"
+        :class="$style['c-spinner-wrapper']">
```
Comment (ashleynolan) [inline packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue]: I would update this to be `c-spinner` and `c-spinner-icon` to show the hierarchy here a little better

File: packages/components/molecules/f-searchbox/src/services/general.services.js [hunk 2]
```diff
@@ -103,9 +103,22 @@ const fullAddressLocalStorageService = {
     }
 };
 
+/**
+ * Extract postcode from address in agreed format
+ * @param {string} address A address string
+ */
+const extractPostcode = address => {
+    try {
+        return address.split(',')[0];
```
Comment (zlatinivanov-je) [inline packages/components/molecules/f-searchbox/src/services/general.services.js:112]: This seems to pick up the first item in the Array, is this always the address format?

File: packages/components/molecules/f-searchbox/src/services/general.services.js [hunk 3]
```diff
@@ -103,9 +103,22 @@ const fullAddressLocalStorageService = {
...
+    } catch (e) {
+        return null;
```
Comment (zlatinivanov-je) [inline packages/components/molecules/f-searchbox/src/services/general.services.js:114]: Maybe worth adding some logging in future.

File: packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        ...mapState('searchbox', [
+            'isFullAddressSearchEnabled',
+            'isLoadingResults'
+        ]),
+
+        /**
+         * Only display loading spinner for Loqate feature for now.
+         *
+         * */
+        shouldDisplayLoadingIndicator () {
```
Comment (dandel10n) [inline packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue:24]: Should this check better be in the parent component? And v-if statement on the component call?

File: packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue [hunk 5]
```diff
@@ -27,11 +27,13 @@
...
                         @input="address = $event.target.value">
                     <span :class="$style['c-search-placeholder']">{{ copy.fieldLabel }}</span>
+
+                    <loading-indicator />
                 </label>
 
                 <cancel-button
                     :class="$style['c-fullAddressFinder-btn']"
-                    @click="onModalClose">
+                    @click.native="onModalClose">
```
Comment (dandel10n) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue]: No need for `.native`

File: packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue [hunk 6]
```diff
@@ -32,11 +32,18 @@
...
-            <div
-                v-if="shouldDisplayLoadingIndicator"
-                :class="$style['c-spinner-wrapper']">
-                <div :class="$style['c-spinner']" />
-            </div>
+            <clear-button
+                v-if="shouldDisplayClearButton"
+                :class="$style['c-search-btn-clear']"
+                data-test-id="search-btn-clear"
+                @click.native="onClearAddress">
```
Comment (dandel10n) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue]: The same here

File: packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue [hunk 7]
```diff
@@ -277,4 +311,58 @@ export default {
...
+
+    svg {
+        width: $icon-size;
+        height: $icon-size;
+        top: 18px;
+        left: 19px;
+        position: absolute;
+
+        g {
+            fill: #535353;
```
Comment (dandel10n) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue]: Can we use colour variable from fozzie-colour-palette here?

File: packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue (hunk 4 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue:24]: I did originally have it there but then im using this component in two places so you'd need the same logic (mapState & computed prop) in both places. Instead when it's here it's just in one place.

**Review by zlatinivanov-je — COMMENTED** [review]
Just one question on the address `extractPostcode`. 
And it seems to beg the question of having a Fozzie spinner/loader component

File: packages/components/molecules/f-searchbox/src/services/general.services.js (hunk 3 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/services/general.services.js:114]: This part of the method shouldn't ever be hit, unless an address isn't passed in. We could get this to `throw` instead in those cases?

File: packages/components/molecules/f-searchbox/src/services/general.services.js (hunk 2 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/services/general.services.js:112]: The address format can be either of these two:

1. `AR511AR {type: string}` - If they continue with a suggested postcode, when they don't want to pick their actual address. 
2. `AR511AR {type: string}, 10 some place, london` - if they pick their address.

In both scenarios they'll end up with just the postcode as it's always a string. :)

File: packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue [hunk 8]
```diff
@@ -1,6 +1,7 @@
 <template>
     <div v-if="isBelowMid">
         <full-address-overlay-search
+            data-test-id="full-address-overlay"
```
Comment (kevinrodrigues) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue:4]: @noorshahJustEat Modal overlay ID.

File: packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchSuggestions.vue [hunk 9]
```diff
@@ -10,12 +10,12 @@
                 $style.item,
                 { [$style.selected]: index === selected }
             ]"
-            data-test-id="suggestion-item"
             @click="getSelectedStreetAddress($event, index, selected)">
             <map-pin-icon :class="$style['c-locationIcon']" />
 
             <div
                 v-if="!selectedStreetLevelAddressId"
+                data-test-id="suggestion-postcode-item"
```
Comment (kevinrodrigues) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchSuggestions.vue:18]: @noorshahJustEat Postcode item

File: packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchSuggestions.vue [hunk 10]
```diff
@@ -26,7 +26,9 @@
                     {{ getDescription(item) }}
                 </p>
             </div>
-            <div v-else>
+            <div
+                data-test-id="suggestion-address-item"
```
Comment (kevinrodrigues) [inline packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchSuggestions.vue]: @noorshahJustEat address item.

File: packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue (hunk 6 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue]: Fixed.

File: packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue (hunk 5 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/components/formElements/FormFullAddressSearchModalOverlay.vue]: Fixed.

File: packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue (hunk 7 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/components/formElements/FormSearchField.vue]: Fixed.

**Review by FayeCarter — APPROVED** [review]

File: packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue (hunk 1 above)
Comment (kevinrodrigues) [reply packages/components/molecules/f-searchbox/src/components/FormStates/FormLoadingIndicator.vue]: Fixed.

**Review by dandel10n — APPROVED** [review]

**Pull request merged**